File: src/schemas/annotation.ts
import { z } from 'zod';

export const PointSignSchema = z.enum(['Positive', 'Negative']);
export type PointSign = z.infer<typeof PointSignSchema>;

export const BoxTypeSchema = z.enum(['Annotate', 'Visual']);
export type BoxType = z.infer<typeof BoxTypeSchema>;

// 前端工具/模式（源自旧版 label-raw）
export const LabelModeSchema = z.enum(['select', 'create', 'delete', 'move']);
export type LabelMode = z.infer<typeof LabelModeSchema>;

export const AnnotationTypeSchema = z.enum(['p_point', 'n_point', 'box']);
export type AnnotationType = z.infer<typeof AnnotationTypeSchema>;

export const AnnotationPointSchema = z.object({
  id: z.string(),
  annotation_id: z.string(),
  x: z.number(),
  y: z.number(),
  sign: PointSignSchema,
});
export type AnnotationPoint = z.infer<typeof AnnotationPointSchema>;

export const AnnotationBoxSchema = z.object({
  id: z.string(),
  annotation_id: z.string(),
  box_type: BoxTypeSchema,
  x1: z.number(),
  y1: z.number(),
  x2: z.number(),
  y2: z.number(),
});
export type AnnotationBox = z.infer<typeof AnnotationBoxSchema>;

export const AnnotationSchema = z.object({
  id: z.string(),
  image_id: z.string(),
  label_id: z.string(),
  boxes: z.array(AnnotationBoxSchema),
  points: z.array(AnnotationPointSchema),
});
export type Annotation = z.infer<typeof AnnotationSchema>;

// Inputs for db_set_annotations
export const PointInputSchema = z.object({
  id: z.string(),
  x: z.number(),
  y: z.number(),
  sign: PointSignSchema,
});
export type PointInput = z.infer<typeof PointInputSchema>;

export const BoxInputSchema = z.object({
  id: z.string(),
  box_type: BoxTypeSchema,
  x1: z.number(),
  y1: z.number(),
  x2: z.number(),
  y2: z.number(),
});
export type BoxInput = z.infer<typeof BoxInputSchema>;

export const AnnotationInputSchema = z.object({
  id: z.string(),
  label_id: z.string(),
  boxes: z.array(BoxInputSchema),
  points: z.array(PointInputSchema),
});
export type AnnotationInput = z.infer<typeof AnnotationInputSchema>;

export const NearestVisualSchema = z.object({
  box_id: z.string(),
  x1: z.number(),
  y1: z.number(),
  x2: z.number(),
  y2: z.number(),
  image_hash: z.string(),
  slice_index: z.number().nullable(),
  image_name: z.string().nullable(),
});
export type NearestVisual = z.infer<typeof NearestVisualSchema>;

export const AnnotationCountSchema = z.object({
  imageHash: z.string(),
  sliceIndex: z.number().nullable(),
  annotationCount: z.number(),
});
export type AnnotationCount = z.infer<typeof AnnotationCountSchema>;

File: crates/db/src/models/annotation.rs
use crate::models::label::Label;

#[derive(Debug, Clone, PartialEq, toasty::Embed, serde::Serialize, serde::Deserialize)]
pub enum BoxType {
    #[column(variant = "annotate")]
    Annotate,
    #[column(variant = "visual")]
    Visual,
}

#[derive(Debug, Clone, PartialEq, toasty::Embed, serde::Serialize, serde::Deserialize)]
pub enum PointSign {
    #[column(variant = "positive")]
    Positive,
    #[column(variant = "negative")]
    Negative,
}

#[derive(Debug, toasty::Model, serde::Serialize)]
pub struct Annotation {
    #[key]
    pub id: String,

    #[index]
    pub image_id: String,

    #[serde(skip)]
    #[belongs_to(key = image_id, references = hash)]
    pub image: toasty::Deferred<crate::models::image::Image>,

    #[index]
    pub label_id: uuid::Uuid,

    #[serde(skip)]
    #[belongs_to(key = label_id, references = id)]
    pub label: toasty::Deferred<Label>,

    #[has_many]
    pub boxes: Vec<AnnotationBox>,

    #[has_many]
    pub points: Vec<AnnotationPoint>,
}

#[derive(Debug, toasty::Model, serde::Serialize)]
pub struct AnnotationBox {
    #[key]
    pub id: String,

    #[index]
    pub annotation_id: String,

    #[serde(skip)]
    #[belongs_to(key = annotation_id, references = id)]
    pub annotation: toasty::Deferred<Annotation>,

    pub box_type: BoxType,

    pub x1: f64,

    pub y1: f64,

    pub x2: f64,

    pub y2: f64,
}

#[derive(Debug, toasty::Model, serde::Serialize)]
pub struct AnnotationPoint {
    #[key]
    pub id: String,

    #[index]
    pub annotation_id: String,

    #[serde(skip)]
    #[belongs_to(key = annotation_id, references = id)]
    pub annotation: toasty::Deferred<Annotation>,

    pub x: f64,

    pub y: f64,

    pub sign: PointSign,
}
